### PR TITLE
meta: get content provider dirs from snap metadata

### PR DESCRIPTION
**Pull request by cmatsuoka**
Obtain the list of directories exposed by content providers listed
in snap.yaml (provided that the corresponding snaps are installed
on the system).

Co-authored-by: Chris Patterson <chris.patterson@canonical.com>
Signed-off-by: Claudio Matsuoka <claudio.matsuoka@canonical.com>

- [x] Have you followed the [guidelines for contributing](https://github.com/snapcore/snapcraft/blob/master/CONTRIBUTING.md)?
- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [ ] Have you successfully run `make lint`?
- [ ] Have you successfully run `pytest tests/unit`?

-----
CRAFT-1246

### CODE REVIEW

**Comment by cmatsuoka** [comment]
Rebased on main.

File: snapcraft/meta/snap_yaml.py [hunk 1]
```diff
@@ -77,14 +81,83 @@ class SnapApp(YamlModel):
...
+        return val
+
+    @property
+    def provider(self) -> Optional[str]:
+        """Return the default content provider name."""
+        if self.default_provider is None:
+            return None
+
+        if ":" in self.default_provider:
+            return self.default_provider.split(":")[0]
```
Comment (mr-cal) [inline snapcraft/meta/snap_yaml.py:120]: pedantic: add a comment about why we grab the data before the colon

File: snapcraft/meta/snap_yaml.py [hunk 2]
```diff
@@ -139,6 +206,59 @@ def unmarshal(cls, data: Dict[str, Any]) -> "SnapMetadata":
 
         return cls(**data)
 
+    def get_provider_content_directories(self) -> List[Path]:
```
Comment (mr-cal) [inline snapcraft/meta/snap_yaml.py:209]: pedantic: some debug or trace-level logs would be nice in these functions.  

I think logging is important in loops when `continue` is called to inform a developer why a section of code was skipped.

File: snapcraft/meta/snap_yaml.py [hunk 3]
```diff
@@ -16,31 +16,35 @@
...
 from snapcraft.utils import get_ld_library_paths, process_version
 
 
-class Socket(YamlModel):
-    """snap.yaml app socket entry."""
-
-    listen_stream: Union[int, str]
-    socket_mode: Optional[int]
-
+class _SnapMetadataModel(YamlModel):
```
Comment (mr-cal) [inline snapcraft/meta/snap_yaml.py:32]: Good idea!

**Review by mr-cal — APPROVED** [review]
Concise and clear, nice work!

File: snapcraft/meta/snap_yaml.py (hunk 2 above)
Comment (cmatsuoka) [reply snapcraft/meta/snap_yaml.py:209]: Will add comments and debug messages as part of the next PR.

File: snapcraft/meta/snap_yaml.py (hunk 1 above)
Comment (cmatsuoka) [reply snapcraft/meta/snap_yaml.py:120]: Will add comments and debug messages as part of the next PR.

**Pull request merged**